README.md
RotRoh v0.1.0
=============

A Go package and command line tool to apply one of several transforms to text input


Features
--------

* Can transform standard input or multiple quoated command line arguments
* Transforms
	* Base64: use the `-base64` option
	* ROT-13: use the `-rot13` option
	* ROT-47: use the `-rot47` option
	* RotRoh: the default if no other options are used
		* RotRoh = ROT-47 + Base64
	* Custom ROT: use the `-custom-rot`
	* Pipeable: use the `-pipe` option to not add a newline to the output
* RotRoh should be pronounced as an homage to Scooby Doo's "Ruh Roh"
* Relies only on the Go stdlib. No 3rd party dependencies.


Examples
--------

### Help

```bash
$ rotroh -h
RotRoh v0.1.0
Usage: rotroh [OPTIONS] STRINGS

OPTIONS:
  -base64          Use Base64 codex (default: false)
  -help            Display this help info (default: false)
  -pipe            Do not output a newline at the end (default: false)
  -rot-custom SET  Use a custom ROT transform set
  -rot13           Use ROT-13 transform (default: false)
  -rot47           Use ROT-47 transform (default: false)
  -rotroh          Use RotRoh codex (default: true)
  -version         Display version info (default: false)

```

### Processing Command Line Arguments

```bash
$ rotroh 'Where is Waldo?'
KDk2QzYgOkQgKDI9NUBu
```

```bash
$ rotroh 'Where is Waldo?' KDk2QzYgOkQgKDI9NUBu 'dzYgPkZERSAzNiA6PyBFOTYgNDI/NUogNDI/NiA3QEM2REVd'
KDk2QzYgOkQgKDI9NUBu
Where is Waldo?
He must be in the candy cane forest.
```

### Processing Standard Input

```bash
$ echo 'Where is Waldo?' | rotroh | rotroh
enM8YSJLKjh+PCI4enN4aH0mcUZyOGxsCg==
```

Don't use `echo` in a pipeline to `rotroh` unless you really want a newline character to be part of the processing


```bash
$ printf 'Where is Waldo?' | rotroh | rotroh
enM8YSJLKjh+PCI4enN4aH0mcUYK
```

Neet to make sure we're not adding extra newlines when using `rotroh` in a pipeline

```bash
$ printf 'Where is Waldo?' | rotroh -pipe | rotroh
Where is Waldo?
```

### Using a Custom ROT

```bash
$ rotroh -rot-custom '1234567890' '1342 Fast Ln.'
0879 Fast Ln.
```

```bash
$ rotroh -pipe -rot-custom '1234567890' '1342 Fast Ln.' | rotroh -rot-custom '1234567890'
1342 Fast Ln.
```


Test Coverage
-------------

73.7% of statements


ToDo
----

* Add custom RotRoh
* Add Base32, etc.
* Allow use of all options for layered and ordered processing
* Allow a config file to store custom transforms, etc.
* data URL codex?


Warning
-------

While this is technically an encryption and obfuscation tool it is not, I repeat **NOT**, to be used were real security is needed. This project is more for fun than security. In other words, it _might_ slow down Velma for a hot second if she couldn't find her glasses. :wink:


Articles & Reference
--------------------

* [ROT-13 Cipher - ROT13 - Online Text Decoder, Encoder, Translator][]
* [ROT-47 Cipher - ROT47 - Online Decoder, Encoder, Translator][]


[ROT-13 Cipher - ROT13 - Online Text Decoder, Encoder, Translator]: https://www.dcode.fr/rot-13-cipher
[ROT-47 Cipher - ROT47 - Online Decoder, Encoder, Translator]: https://www.dcode.fr/rot-47-cipher



CHANGELOG.md
CHANGELOG
=========


v0.1.0
------

* Initial release
* Base64 encoding and decoding (auto detects what is needed)
* ROT-13 transform
* ROT-47 transform
* Custom ROT transform
* RotRoh encoding and decoding (ROT-47 + Base64)
* Pipeline option
* Unit testing with coverage



rotroh_test.go
package rotroh

import (
	"testing"
)

func TestBase64String(t *testing.T) {
	want := "VXNlck5hbWU6UGFzc3dvcmQ="
	got, err := Base64String("UserName:Password")
	if err != nil {
		t.Fatalf("error: %s", err.Error())
	}
	if got != want {
		t.Fatalf("expected: %q, got: %q", want, got)
	}
}

func TestRot13String(t *testing.T) {
	want := "Uv Zn!"
	got := Rot13String("Hi Ma!")
	if got != want {
		t.Fatalf("expected: %q, got: %q", want, got)
	}
}

func TestRot47String(t *testing.T) {
	want := "(96C6 :D (2=5@n"
	got := Rot47String("Where is Waldo?")
	if got != want {
		t.Fatalf("expected: %q, got: %q", want, got)
	}
}

func TestRotRoh47String(t *testing.T) {
	want := "KDk2QzYgOkQgKDI9NUBu"
	got, err := RotRoh47String("Where is Waldo?")
	if err != nil {
		t.Fatalf("error: %s", err.Error())
	}
	if got != want {
		t.Fatalf("expected: %q, got: %q", want, got)
	}
}

func TestRotCustomString(t *testing.T) {
	want := "0879 Fast Ln."
	got, err := RotCustomString("1342 Fast Ln.", "1234567890")
	if err != nil {
		t.Fatalf("error: %s", err.Error())
	}
	if got != want {
		t.Fatalf("expected: %q, got: %q", want, got)
	}
}


rotroh.go
package rotroh

import (
	"encoding/base64"
	"fmt"
	"strings"
)

const (
	rot13Src = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
	rot13Des = "NOPQRSTUVWXYZABCDEFGHIJKLMnopqrstuvwxyzabcdefghijklm"
	rot47Src = `!"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_` + "`abcdefghijklmnopqrstuvwxyz{|}~"
	rot47Des = "~}|{zyxwvutsrqponmlkjihgfedcba`" + `_^]\[ZYXWVUTSRQPONMLKJIHGFEDCBA@?>=<;:9876543210/.-,+*)('&%$#"!`
)

// Base64String does a Base64 encoding or a Base64 decoding
func Base64String(input string) (result string, err error) {
	if IsBase64String(input) {
		// log.Println("Input is Base64 encoded")
		b, err := base64.StdEncoding.DecodeString(input)
		if err == nil {
			result = string(b)
		}
		if err != nil {
			return "", err
		}
	} else {
		// log.Println("Input is not Base64 encoded")
		result = base64.StdEncoding.EncodeToString([]byte(input))
	}

	return result, err
}

// Reverse returns its argument string reversed rune-wise left to right.
// @see https://github.com/golang/example/blob/master/stringutil/reverse.go
func reverse(s string) string {
	r := []rune(s)
	for i, j := 0, len(r)-1; i < len(r)/2; i, j = i+1, j-1 {
		r[i], r[j] = r[j], r[i]
	}
	return string(r)
}

// RotCustomString allows for a custom characterset based ROT
func RotCustomString(input, source string) (result string, err error) {
	reversed := reverse(source)
	if len(source)%2 != 0 {
		err = fmt.Errorf("error: the source must be an even number of characters")
		return "", err
	}

	for _, r := range input {
		s := string(r)
		i := strings.Index(source, s)
		// log.Printf("RotCustom() | s: %q | i: %d\n", s, i)
		if i > -1 {
			result += string(reversed[i])
		} else {
			result += s
		}
	}

	return result, err
}

// Rot13String does a ROT-13 transform on a string
func Rot13String(input string) string {
	result := ""
	for _, r := range input {
		s := string(r)
		i := strings.Index(rot13Src, s)
		// log.Printf("Rot13String() | s: %q | i: %d\n", s, i)
		if i > -1 {
			result += string(rot13Des[i])
		} else {
			result += s
		}
	}
	return result
}

// Rot47String does a ROT-47 transform on a string
func Rot47String(input string) string {
	result := ""
	for _, r := range input {
		i := int(r)
		x := i
		if 32 < i && i < 80 {
			x = i + 47
		} else if 79 < i && i < 127 {
			x = i - 47
		}
		s := string(rune(x))
		// log.Printf("Rot47String() | r: %q | i: %3d | x: %3d | s: %q\n", r, i, x, s)
		result += s
	}
	return result
}

// IsBase64String returns true if the input was a already base64 encoded
func IsBase64String(input string) bool {
	decoded, err := base64.StdEncoding.DecodeString(input)
	if err != nil {
		// log.Printf("IsBase64String(%s) | Error: %s\n", input, err.Error())
		return false
	}
	encoded := base64.StdEncoding.EncodeToString([]byte(decoded))
	if encoded == input {
		return true
	}
	return false
}

// RotRoh47String does a ROT-47 transform then Base64 encoding or a Base64 decoding then a ROT-47 transform
func RotRoh47String(input string) (result string, err error) {
	if IsBase64String(input) {
		// log.Println("Input is Base64 encoded")
		decoded, err := base64.StdEncoding.DecodeString(input)
		if err != nil {
			return result, err
		}
		result = Rot47String(string(decoded))
	} else {
		// log.Println("Input is not Base64 encoded")
		msg := Rot47String(input)
		result = base64.StdEncoding.EncodeToString([]byte(msg))
	}

	return result, err
}


cmd/rotroh/main.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"

	"github.com/runeimp/rotroh"
)

const (
	appName    = "RotRoh"
	appVersion = "0.1.0"
	usage      = `Usage: %s [OPTIONS] STRINGS

OPTIONS:
`
)

var appLabel = fmt.Sprintf("%s v%s", appName, appVersion)

func main() {
	base64Ptr := flag.Bool("base64", false, "Use Base64 codex")
	helpPtr := flag.Bool("help", false, "Display this help info")
	pipeModePtr := flag.Bool("pipe", false, "Do not output a newline at the end")
	rot13Ptr := flag.Bool("rot13", false, "Use ROT-13 transform")
	rot47Ptr := flag.Bool("rot47", false, "Use ROT-47 transform")
	rotCustomPtr := flag.String("rot-custom", "", "Use a custom ROT transform set")
	rotRohPtr := flag.Bool("rotroh", true, "Use RotRoh codex")
	versionPtr := flag.Bool("version", false, "Display version info")

	noRotRoh := false

	flag.Usage = func() {
		fmt.Println(appLabel)
		fmt.Fprintf(flag.CommandLine.Output(), usage, filepath.Base(os.Args[0]))

		flag.VisitAll(func(f *flag.Flag) {
			optionName := fmt.Sprintf("-%s", f.Name)
			if len(f.DefValue) > 0 {
				fmt.Fprintf(flag.CommandLine.Output(), "  %-15s  %s (default: %v)\n", optionName, f.Usage, f.DefValue)
			} else {
				fmt.Fprintf(flag.CommandLine.Output(), "  %-15s  %s\n", optionName+" SET", f.Usage)
			}
		})
		fmt.Println()
	}

	flag.Parse()

	if *helpPtr {
		flag.Usage()
		os.Exit(0)
	}

	if *versionPtr {
		fmt.Println(appLabel)
		os.Exit(0)
	}

	args := flag.Args()

	stat, _ := os.Stdin.Stat()
	if (stat.Mode() & os.ModeCharDevice) == 0 {
		b, _ := ioutil.ReadAll(os.Stdin)
		args = []string{string(b)}
	}

	for _, arg := range args {
		if *rot13Ptr && *rot47Ptr {
			fmt.Fprintf(os.Stderr, "You can only use one transform of either ROT-13, ROT-47, Base64, or RotRoh\n")
			os.Exit(1)
		}
		// fmt.Printf("arg: %q\n", arg)

		var err error
		result := arg

		if *rot47Ptr {
			// fmt.Println("ROT-47")
			result = rotroh.Rot47String(result)
			noRotRoh = true
		} else if *rot13Ptr {
			// fmt.Println("ROT-13")
			result = rotroh.Rot13String(result)
			noRotRoh = true
		} else if *rotCustomPtr != "" {
			// fmt.Println("RotCustom")
			result, err = rotroh.RotCustomString(result, *rotCustomPtr)
			noRotRoh = true
		} else if *base64Ptr {
			// fmt.Println("Base64")
			result, err = rotroh.Base64String(result)
			noRotRoh = true
		}

		if *rotRohPtr && noRotRoh == false {
			// fmt.Println("RotRoh")
			result, err = rotroh.RotRoh47String(result)
		}

		if err != nil {
			fmt.Println(err.Error())
		}
		if *pipeModePtr {
			fmt.Print(result)
		} else {
			fmt.Println(result)
		}
	}
}

func test() {
	result := rotroh.Rot13String("Hi Ma!")
	log.Printf("main() | result: %q\n", result)

	result = rotroh.Rot13String(result)
	log.Printf("main() | result: %q\n", result)

	result = rotroh.Rot47String("Ahegao!")
	log.Printf("main() | result: %q\n", result)

	result = rotroh.Rot47String(result)
	log.Printf("main() | result: %q\n", result)

	result, err := rotroh.RotRoh47String("Ahegao!")
	if err != nil {
		log.Printf("main() | error: %q\n", err.Error())
	} else {
		log.Printf("main() | result: %q\n", result)
	}

	result, err = rotroh.RotRoh47String(result)
	if err != nil {
		log.Printf("main() | error: %q\n", err.Error())
	} else {
		log.Printf("main() | result: %q\n", result)
	}
}
